File: paleo_workbench/mapping_workspace/source_usage.py
"""Version → mapping usage reverse query (V13 W-I / W-M).

回答「这个 DataVersion 被哪些地图内容引用？」——从既有权威**派生投影**，
不新建任何存储：

- ``MappingWorkspaceState.memberships``：图层级绑定（``source_version_id``，
  V13 起含 ``source_asset_id``）；
- ``ProjectDocument.factor_map_tasks``：factor 任务的产品格网版本
  （``grid_artifact_version_id``）——factor 子图层经任务间接绑定；
- ``ProjectDocument.compilation_input_sets``：综合编图输入集的钉定版本
  （``pinned_version_id`` / ``resolved_asset_id``）；
- ``ProjectDocument.map_products``：成图产品的输出版本与 run 输入；
- 可选 catalog：消费该版本的 run（有界）。

正向（图层→版本）由 membership 记录本身回答；本模块补齐反向闭环。
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import TYPE_CHECKING, Any

if TYPE_CHECKING:  # pragma: no cover - typing only
    from paleo_workbench.mapping_workspace.stage_state import MappingWorkspaceState
    from paleo_workbench.project.models import ProjectDocument

#: run 级用途的硬上限（run 输入可能非常宽，如 53 井预测）。
MAX_RUN_USAGE_ENTRIES = 100

# 用途类别词汇（稳定字符串，UI/Agent 直接消费）。
USAGE_LAYER = "layer"
USAGE_FACTOR_GRID = "factor_grid"
USAGE_COMPILATION_INPUT = "compilation_input"
USAGE_MAP_PRODUCT_OUTPUT = "map_product_output"
USAGE_MAP_PRODUCT_INPUT = "map_product_input"
USAGE_RUN_INPUT = "run_input"


@dataclass(frozen=True)
class VersionUsage:
    """一条「谁在用这个版本」的记录（纯派生，不持久化）。"""

    kind: str
    ref_id: str
    label: str
    #: 匹配到的版本 id（资产级查询时标明是哪个版本被引用）。
    version_id: str = ""
    stage: str = ""
    role: str = ""
    status: str = ""


@dataclass(frozen=True)
class VersionUsageReport:
    usages: list[VersionUsage] = field(default_factory=list)
    truncated: bool = False


def usages_of_version(
    version_id: str,
    *,
    workspace: "MappingWorkspaceState | None",
    project: "ProjectDocument | None",
    catalog: Any = None,
    include_runs: bool = True,
) -> VersionUsageReport:
    """某 DataVersion 的全部地图侧引用（有界，truncated 标志如实）。"""
    vid = str(version_id or "")
    if not vid:
        return VersionUsageReport()
    usages: list[VersionUsage] = []
    truncated = False
    if workspace is not None:
        for layer_id, record in workspace.memberships.items():
            if record.source_version_id == vid:
                usages.append(VersionUsage(
                    kind=USAGE_LAYER,
                    ref_id=str(layer_id),
                    label=f"图层 {layer_id}",
                    version_id=vid,
                    stage=record.created_stage,
                    role=record.role.value,
                ))
    if project is not None:
        for task in getattr(project, "factor_map_tasks", None) or []:
            if str(getattr(task, "grid_artifact_version_id", "") or "") == vid:
                usages.append(VersionUsage(
                    kind=USAGE_FACTOR_GRID,
                    ref_id=str(task.id),
                    label=str(getattr(task, "name", "") or task.id),
                    version_id=vid,
                    role="factor_grid",
                ))
        for raw in getattr(project, "compilation_input_sets", None) or []:
            if not isinstance(raw, dict):
                continue
            set_id = str(raw.get("id") or "")
            set_name = str(raw.get("name") or set_id)
            frozen = bool(raw.get("frozen"))
            for entry in raw.get("entries") or []:
                if not isinstance(entry, dict):
                    continue
                if str(entry.get("pinned_version_id") or "") == vid:
                    usages.append(VersionUsage(
                        kind=USAGE_COMPILATION_INPUT,
                        ref_id=set_id,
                        label=set_name,
                        version_id=vid,
                        status="frozen" if frozen else "draft",
                    ))
        for record in getattr(project, "map_products", None) or []:
            output_vid = str(getattr(record, "output_version_id", "") or "")
            if output_vid == vid:
                usages.append(VersionUsage(
                    kind=USAGE_MAP_PRODUCT_OUTPUT,
                    ref_id=str(record.id),
                    label=str(getattr(record, "product_name", "") or record.id),
                    version_id=vid,
                    status=str(getattr(record, "lifecycle", "") or ""),
                ))
                continue
            run_id = str(getattr(record, "run_id", "") or "")
            if run_id and catalog is not None:
                if _run_consumes(catalog, run_id, vid):
                    usages.append(VersionUsage(
                        kind=USAGE_MAP_PRODUCT_INPUT,
                        ref_id=str(record.id),
                        label=str(getattr(record, "product_name", "") or record.id),
                        version_id=vid,
                        status=str(getattr(record, "lifecycle", "") or ""),
                    ))
    if include_runs and catalog is not None:
        runs = _runs_consuming(catalog, vid)
        for index, run in enumerate(runs):
            if index >= MAX_RUN_USAGE_ENTRIES:
                truncated = True
                break
            usages.append(VersionUsage(
                kind=USAGE_RUN_INPUT,
                ref_id=str(run.id),
                label=f"run {run.operation}",
                version_id=vid,
                status=str(run.status or ""),
            ))
    return VersionUsageReport(usages=usages, truncated=truncated)


def usages_of_asset(
    asset_id: str,
    *,
    workspace: "MappingWorkspaceState | None",
    project: "ProjectDocument | None",
    catalog: Any = None,
) -> VersionUsageReport:
    """某 DataAsset 的全部地图侧引用（聚合其各版本 + 资产级直绑）。

    有 catalog：遍历资产全部版本逐一反查（版本数通常个位数到几十）。
    无 catalog：只回答资产级绑定（membership.source_asset_id、输入集
    resolved_asset_id）——诚实降级，不伪造版本级结论。
    """
    aid = str(asset_id or "")
    if not aid:
        return VersionUsageReport()
    usages: list[VersionUsage] = []
    seen: set[tuple[str, str, str]] = set()
    if catalog is not None:
        versions = _versions_of_asset(catalog, aid)
        for version in versions:
            report = usages_of_version(
                version.id, workspace=workspace, project=project,
                catalog=catalog, include_runs=False,
            )
            for usage in report.usages:
                key = (usage.kind, usage.ref_id, usage.version_id)
                if key not in seen:
                    seen.add(key)
                    usages.append(usage)
    if workspace is not None:
        for layer_id, record in workspace.memberships.items():
            if record.source_asset_id == aid:
                key = (USAGE_LAYER, str(layer_id), record.source_version_id)
                if key not in seen:
                    seen.add(key)
                    usages.append(VersionUsage(
                        kind=USAGE_LAYER,
                        ref_id=str(layer_id),
                        label=f"图层 {layer_id}",
                        version_id=record.source_version_id,
                        stage=record.created_stage,
                        role=record.role.value,
                    ))
    if project is not None:
        for raw in getattr(project, "compilation_input_sets", None) or []:
            if not isinstance(raw, dict):
                continue
            for entry in raw.get("entries") or []:
                if (isinstance(entry, dict)
                        and str(entry.get("resolved_asset_id") or "") == aid):
                    set_id = str(raw.get("id") or "")
                    key = (USAGE_COMPILATION_INPUT, set_id,
                           str(entry.get("pinned_version_id") or ""))
                    if key not in seen:
                        seen.add(key)
                        usages.append(VersionUsage(
                            kind=USAGE_COMPILATION_INPUT,
                            ref_id=set_id,
                            label=str(raw.get("name") or set_id),
                            version_id=str(entry.get("pinned_version_id") or ""),
                            status="frozen" if raw.get("frozen") else "draft",
                        ))
    return VersionUsageReport(usages=usages, truncated=False)


def usage_counts(report: VersionUsageReport) -> dict[str, int]:
    """按类别计数（状态栏/Inspector 摘要用）。"""
    counts: dict[str, int] = {}
    for usage in report.usages:
        counts[usage.kind] = counts.get(usage.kind, 0) + 1
    return counts


# ---------------------------------------------------------------- helpers
def _run_consumes(catalog: Any, run_id: str, version_id: str) -> bool:
    try:
        run = catalog.get_run(run_id)
    except Exception:
        return False
    if run is None:
        return False
    return version_id in list(run.input_version_ids or [])


def _runs_consuming(catalog: Any, version_id: str) -> list[Any]:
    try:
        runs = catalog.runs_consuming(version_id=version_id)
    except Exception:
        return []
    return list(runs or [])


def _versions_of_asset(catalog: Any, asset_id: str) -> list[Any]:
    try:
        maps = catalog._ensure_maps()
        return list(maps.versions_by_asset.get(asset_id, ()))
    except Exception:
        return []

File: libs/workspace/src/source_usage.cpp
// V14 source usage reverse query — port of source_usage.py (see
// source_usage.hpp).
#include "pwb/workspace/source_usage.hpp"

#include <tuple>

namespace pwb::workspace {

namespace {

// str(value or ""): absent/null/"" -> ""; strings kept; other scalars
// stringified via dump (closest honest mapping of Python str()).
std::string str_or_empty(const domain::Json& parent, const char* key) {
    if (!parent.is_object()) return "";
    auto it = parent.find(key);
    if (it == parent.end() || it->is_null()) return "";
    if (it->is_string()) return it->get<std::string>();
    if (it->is_number_integer()) return std::to_string(it->get<long long>());
    if (it->is_number_float()) return it->dump();
    if (it->is_boolean()) return it->get<bool>() ? "True" : "False";
    return "";
}

bool run_consumes(const UsageCatalog& catalog, const std::string& run_id,
                  const std::string& version_id) {
    const auto inputs = catalog.run_input_version_ids(run_id);
    if (!inputs.has_value()) return false;
    return std::find(inputs->begin(), inputs->end(), version_id) !=
           inputs->end();
}

std::vector<RunSummary> runs_consuming(const UsageCatalog& catalog,
                                       const std::string& version_id) {
    try {
        return catalog.runs_consuming(version_id);
    } catch (...) {
        return {};  // Python: except Exception -> []
    }
}

std::vector<std::string> versions_of_asset(const UsageCatalog& catalog,
                                           const std::string& asset_id) {
    try {
        return catalog.versions_of_asset(asset_id);
    } catch (...) {
        return {};  // Python: except Exception -> []
    }
}

}  // namespace

VersionUsageReport usages_of_version(
    const std::string& version_id, const MappingWorkspaceState* workspace,
    const domain::Json* project_root, const UsageCatalog* catalog,
    bool include_runs) {
    VersionUsageReport report;
    const std::string vid = version_id;
    if (vid.empty()) return report;
    auto& usages = report.usages;
    if (workspace != nullptr) {
        for (const auto& [layer_id, record] : workspace->memberships) {
            if (record.source_version_id != vid) continue;
            VersionUsage usage;
            usage.kind = std::string(kUsageLayer);
            usage.ref_id = layer_id;
            usage.label = "图层 " + layer_id;
            usage.version_id = vid;
            usage.stage = record.created_stage;
            usage.role = record.role;
            usages.push_back(std::move(usage));
        }
    }
    if (project_root != nullptr && project_root->is_object()) {
        if (project_root->contains("factor_map_tasks") &&
            (*project_root)["factor_map_tasks"].is_array()) {
            for (const auto& task : (*project_root)["factor_map_tasks"]) {
                if (str_or_empty(task, "grid_artifact_version_id") != vid) {
                    continue;
                }
                VersionUsage usage;
                usage.kind = std::string(kUsageFactorGrid);
                usage.ref_id = str_or_empty(task, "id");
                const std::string name = str_or_empty(task, "name");
                usage.label = name.empty() ? usage.ref_id : name;
                usage.version_id = vid;
                usage.role = "factor_grid";
                usages.push_back(std::move(usage));
            }
        }
        if (project_root->contains("compilation_input_sets") &&
            (*project_root)["compilation_input_sets"].is_array()) {
            for (const auto& raw : (*project_root)["compilation_input_sets"]) {
                if (!raw.is_object()) continue;
                const std::string set_id = str_or_empty(raw, "id");
                const std::string set_name = str_or_empty(raw, "name");
                const bool frozen = raw.value("frozen", false);
                if (!raw.contains("entries") || !raw["entries"].is_array()) {
                    continue;
                }
                for (const auto& entry : raw["entries"]) {
                    if (!entry.is_object()) continue;
                    if (str_or_empty(entry, "pinned_version_id") != vid) {
                        continue;
                    }
                    VersionUsage usage;
                    usage.kind = std::string(kUsageCompilationInput);
                    usage.ref_id = set_id;
                    usage.label = set_name.empty() ? set_id : set_name;
                    usage.version_id = vid;
                    usage.status = frozen ? "frozen" : "draft";
                    usages.push_back(std::move(usage));
                }
            }
        }
        if (project_root->contains("map_products") &&
            (*project_root)["map_products"].is_array()) {
            for (const auto& record : (*project_root)["map_products"]) {
                const std::string output_vid =
                    str_or_empty(record, "output_version_id");
                const std::string ref_id = str_or_empty(record, "id");
                const std::string product_name =
                    str_or_empty(record, "product_name");
                const std::string label =
                    product_name.empty() ? ref_id : product_name;
                const std::string lifecycle =
                    str_or_empty(record, "lifecycle");
                if (output_vid == vid) {
                    VersionUsage usage;
                    usage.kind = std::string(kUsageMapProductOutput);
                    usage.ref_id = ref_id;
                    usage.label = label;
                    usage.version_id = vid;
                    usage.status = lifecycle;
                    usages.push_back(std::move(usage));
                    continue;
                }
                const std::string run_id = str_or_empty(record, "run_id");
                if (!run_id.empty() && catalog != nullptr &&
                    run_consumes(*catalog, run_id, vid)) {
                    VersionUsage usage;
                    usage.kind = std::string(kUsageMapProductInput);
                    usage.ref_id = ref_id;
                    usage.label = label;
                    usage.version_id = vid;
                    usage.status = lifecycle;
                    usages.push_back(std::move(usage));
                }
            }
        }
    }
    if (include_runs && catalog != nullptr) {
        const std::vector<RunSummary> runs =
            runs_consuming(*catalog, vid);
        for (std::size_t index = 0; index < runs.size(); ++index) {
            if (index >= kMaxRunUsageEntries) {
                report.truncated = true;
                break;
            }
            VersionUsage usage;
            usage.kind = std::string(kUsageRunInput);
            usage.ref_id = runs[index].id;
            usage.label = "run " + runs[index].operation;
            usage.version_id = vid;
            usage.status = runs[index].status;
            usages.push_back(std::move(usage));
        }
    }
    return report;
}

VersionUsageReport usages_of_asset(const std::string& asset_id,
                                   const MappingWorkspaceState* workspace,
                                   const domain::Json* project_root,
                                   const UsageCatalog* catalog) {
    VersionUsageReport report;
    const std::string aid = asset_id;
    if (aid.empty()) return report;
    auto& usages = report.usages;
    std::vector<std::tuple<std::string, std::string, std::string>> seen;
    auto already_seen =
        [&seen](const std::string& kind, const std::string& ref_id,
                const std::string& version_id) {
            const std::tuple<std::string, std::string, std::string> key(
                kind, ref_id, version_id);
            if (std::find(seen.begin(), seen.end(), key) != seen.end()) {
                return true;
            }
            seen.push_back(key);
            return false;
        };
    if (catalog != nullptr) {
        // Versions of an asset are usually single-digit to a few dozen.
        for (const std::string& version_id :
             versions_of_asset(*catalog, aid)) {
            VersionUsageReport per_version = usages_of_version(
                version_id, workspace, project_root, catalog, false);
            for (auto& usage : per_version.usages) {
                if (!already_seen(usage.kind, usage.ref_id,
                                  usage.version_id)) {
                    usages.push_back(std::move(usage));
                }
            }
        }
    }
    if (workspace != nullptr) {
        for (const auto& [layer_id, record] : workspace->memberships) {
            if (record.source_asset_id != aid) continue;
            if (already_seen(std::string(kUsageLayer), layer_id,
                             record.source_version_id)) {
                continue;
            }
            VersionUsage usage;
            usage.kind = std::string(kUsageLayer);
            usage.ref_id = layer_id;
            usage.label = "图层 " + layer_id;
            usage.version_id = record.source_version_id;
            usage.stage = record.created_stage;
            usage.role = record.role;
            usages.push_back(std::move(usage));
        }
    }
    if (project_root != nullptr && project_root->is_object() &&
        project_root->contains("compilation_input_sets") &&
        (*project_root)["compilation_input_sets"].is_array()) {
        for (const auto& raw : (*project_root)["compilation_input_sets"]) {
            if (!raw.is_object()) continue;
            if (!raw.contains("entries") || !raw["entries"].is_array()) {
                continue;
            }
            for (const auto& entry : raw["entries"]) {
                if (!entry.is_object()) continue;
                if (str_or_empty(entry, "resolved_asset_id") != aid) {
                    continue;
                }
                const std::string set_id = str_or_empty(raw, "id");
                const std::string pinned =
                    str_or_empty(entry, "pinned_version_id");
                if (already_seen(std::string(kUsageCompilationInput),
                                 set_id, pinned)) {
                    continue;
                }
                VersionUsage usage;
                usage.kind = std::string(kUsageCompilationInput);
                usage.ref_id = set_id;
                const std::string set_name = str_or_empty(raw, "name");
                usage.label = set_name.empty() ? set_id : set_name;
                usage.version_id = pinned;
                usage.status =
                    raw.value("frozen", false) ? "frozen" : "draft";
                usages.push_back(std::move(usage));
            }
        }
    }
    return report;
}

std::map<std::string, std::size_t> usage_counts(
    const VersionUsageReport& report) {
    std::map<std::string, std::size_t> counts;
    for (const VersionUsage& usage : report.usages) {
        ++counts[usage.kind];
    }
    return counts;
}

}  // namespace pwb::workspace
